### ISSUE TITLE
[usingfastindex] Indexes on AUTO_INCREMENT column should not be selected for fast index creation

### ISSUE THREAD

**Issue by davidducos** [maxbube/mydumper#322]
Hi @davidducos,
I gave it a go with usingFastIndex and am getting some error messages:

`2021-04-27 04:17:10 [INFO] - Fast index creation will be use for table: 35k_020_cba_perfhw_3681_vm_mysql_3326.job_definition_snapshot
2021-04-27 04:17:10 [ERROR] - Error restoring 35k_020_cba_perfhw_3681_vm_mysql_3326.job_definition_snapshot from file 35k_cba_perfhw_3681_vm_20_mysql_3326.job_definition_snapshot-schema.sql.gz: Incorrect table definition; there can be only one auto column and it must be d
efined as a key
```
CREATE TABLE `job_definition_snapshot` (
  `seq` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `hash` binary(20) NOT NULL,
  `iv` varchar(30) NOT NULL,
  `data` blob NOT NULL,
  PRIMARY KEY (`hash`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;`

```
_Originally posted by @VMEMachuca in https://github.com/maxbube/mydumper/issues/284#issuecomment-827553482_

**Comment by davidducos** [comment]
Hi @VMEMachuca,
Can you execute SHOW `CREATE TABLE 35k_020_cba_perfhw_3681_vm_mysql_3326.job_definition_snapshot;` ?

**Comment by VMEMachuca** [comment]
Here is from the source database:

| job_definition_snapshot | CREATE TABLE `job_definition_snapshot` (
  `seq` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `hash` binary(20) NOT NULL,
  `iv` varchar(30) NOT NULL,
  `data` blob NOT NULL,
  PRIMARY KEY (`hash`),
  KEY `IDX_JOB_DEFINITION_SNAPSHOT_SEQ` (`seq`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 |


From: davidducos ***@***.***>
Date: Tuesday, April 27, 2021 at 5:15 AM
To: maxbube/mydumper ***@***.***>
Cc: Victor Machuca ***@***.***>, Mention ***@***.***>
Subject: [External Sender] Re: [maxbube/mydumper] [usingfastindex] Error creating table (#322)

Hi @VMEMachuca<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_VMEMachuca&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=ZqVsHZ7yztTDpeSYnh4qE7DiT3NbKx2e7gB3wFmhWB0&m=cdPtNodAvvkvaFqEsp-9FKeEB96s0IpH5q_0MXz9s2w&s=ryEGSFHFItrrRqh5WrO-9Q9uUXXtzLCaeeyLhRLI3I0&e=>,
Can you execute SHOW CREATE TABLE 35k_020_cba_perfhw_3681_vm_mysql_3326.job_definition_snapshot; ?

**Comment by davidducos** [comment]
The issue is that we are not filtering out indexes over the auto_increment column. This is a confirmed bug, I will review and add a patch before merging usingfastindex to master.

**Comment by davidducos** [comment]
I added a fix, that ignores the index if the first column of the index is the auto_increment column. I used your tables for testing and now, it works.
Please, pull latest version of usingFastIndex and test it again, it should be working now.

**Comment by druud** [comment]
Looks like a good fix to me.

As an alternative, I'll also benchmark disabling the autoinc temporarily, to see how that pans out.
(as the autoinc should not be needed during the data-load phase)

**Comment by druud** [comment]
JFYI: adding back indexes, can in many cases run in the background. All indexes are best added back in a single alter statement.

**Comment by davidducos** [comment]
Hi @druud, we are adding the indexes in a single alter when possible.

**Comment by davidducos** [comment]
@druud how you propose to disable the auto_increment? modifying the column to remove it and then perform an alter to add it?

**Comment by druud** [comment]
Not remove the column, only remove the autoinc attribute.

**Comment by druud** [comment]
I also experimented with PARTITION EXCHANGE, before adding back the indexes. Then all partitions can add indexes in parallel. But that only works for MySQL, not for MariaDB, as that doesn't have the mode to exchange partitions back without row-checks.

**Comment by druud** [comment]
My tables generally have hundreds of millions of rows, so then it can go many hours faster. :)

**Comment by druud** [comment]
For PARTITION trickery, the partitioning needs to be either on HASH() (which is MOD()), or on RANGE (not RANGE COLUMNS).

**Comment by davidducos** [comment]
@druud is any feature request here? if it is, please, open a new ticket and I will be please to prioritize and implement it. 

Would be nice to see how we can implement the partition exchange.
